lib/pacemaker/pcmk_sched_group.c

```diff
@@ -228,10 +228,26 @@ member_internal_constraints(gpointer data, gpointer user_data)
 
     } else {
         // Order this member relative to the previous one
+
         pcmk__order_starts(member_data->previous_member, member,
                            pe_order_implies_then|pe_order_runnable_left);
         pcmk__order_stops(member, member_data->previous_member,
                           pe_order_optional|pe_order_restart);
+
+        /* In unusual circumstances (such as adding a new member to the middle
+         * of a group with unmanaged later members), this member may be active
+         * while the previous (new) member is inactive. In this situation, the
+         * usual restart orderings will be irrelevant, so we need to order this
+         * member's stop before the previous member's start.
+         */
+        if ((member->running_on != NULL)
```

Comment:
Seems reasonable. I haven't thought of a way to avoid the explicit special case "running_on" code -- granted, my knowledge of constraints is "meh," but this might be the best we can do.

FYI, explicit constraints (without grouping) are subject to the same old behavior but without an invalid transition:
```
    <constraints>
      <rsc_order first="grp1a" then="interloper" first-action="start" then-action="start" id="order-grp1a-interloper-mandatory"/>
      <rsc_order first="interloper" then="grp1b" first-action="start" then-action="start" id="order-interloper-grp1b-mandatory"/>
      <rsc_order first="grp1b" then="grp1c" first-action="start" then-action="start" id="order-grp1b-grp1c-mandatory"/>
      <rsc_colocation rsc="interloper" with-rsc="grp1a" score="INFINITY" id="colocation-interloper-grp1a-INFINITY"/>
      <rsc_colocation rsc="grp1b" with-rsc="interloper" score="INFINITY" id="colocation-grp1b-interloper-INFINITY"/>
      <rsc_colocation rsc="grp1c" with-rsc="grp1b" score="INFINITY" id="colocation-grp1c-grp1b-INFINITY"/>
    </constraints>

...
# crm_simulate ...
interloper	(ocf:pacemaker:Dummy):	 Started rhel8-2
grp1b	(ocf:pacemaker:Dummy):	 Started rhel8-2 (unmanaged)
grp1c	(ocf:pacemaker:Dummy):	 Started rhel8-2 (unmanaged)
```

Reply:
I've got more corner cases left to handle, some of which involve the equivalent explicit constraints. Just leaving those aside for now.

I think it would be a good idea to stop treating groups as a resource variant, and instead remap them internally to explicit constraints, so they always behave identically. But that would be too big a can a worms to open anytime soon.